File: 백준/Silver/9012. 괄호/괄호.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');

let num = Number(input.shift());


for(let i=0; i<num; i++) {
    const cases = input[i];
    let answer = 'YES'
    const stack = [];
    for(let j=0; j<cases.length; j++) {
        if(cases[j] === '(') {
            stack.push(cases[j]);
        } else if(cases[j] === ')') {
            if(stack.length === 0) {
                answer = 'NO'
                break;
            }
            stack.pop();
        }
    }
    if(stack.length !==0) {
        answer = 'NO'
    }
    console.log(answer);
}

File: 백준/Silver/9012. 괄호/괄호.py
n = int(input())

for _ in range(n):
    a = input()
    s = list(a)
    tmp = 0
    for i in s:
        if i == '(':
            tmp += 1
        if i == ')':
            tmp -= 1
        if tmp < 0:
            print('NO')
            break
    if tmp > 0:
        print('NO')
    if tmp == 0:
        print('YES')